File: docs/prd/forum-app/_capture.mjs
import { chromium } from 'playwright';
import { mkdir, writeFile } from 'node:fs/promises';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = dirname(fileURLToPath(import.meta.url));
const OUT = join(ROOT, 'screenshots');
const BASE = 'http://127.0.0.1:5173/';

await mkdir(OUT, { recursive: true });

const browser = await chromium.launch({ headless: true, channel: 'chrome' });
const shots = [];

async function shot(page, name, locator) {
  const file = `${name}.png`;
  const path = join(OUT, file);
  try {
    if (locator) {
      const el = locator.first();
      if (await el.isVisible({ timeout: 2500 })) {
        await el.screenshot({ path, timeout: 8000 });
        shots.push(file);
        console.log('ok', file);
        return;
      }
    }
  } catch (err) {
    console.log('fallback', file, String(err.message || err).slice(0, 80));
  }
  await page.screenshot({ path, fullPage: false });
  shots.push(file);
  console.log('page', file);
}

async function goto(page, hash, waitMs = 900) {
  await page.goto(`${BASE}${hash}`, { waitUntil: 'networkidle' });
  await page.waitForTimeout(waitMs);
}

const admin = await browser.newContext({ viewport: { width: 1440, height: 900 } });
const ap = await admin.newPage();

await goto(ap, '#/forum/forum-overview');
await shot(ap, '后台-概览-页面整体');
await shot(ap, '论坛应用-左侧菜单', ap.locator('.ant-layout-sider').first());
await shot(ap, '后台-概览-KPI指标卡', ap.locator('.overview-kpi-grid').first());
await shot(ap, '后台-概览-日期筛选', ap.locator('.overview-page .ant-picker, .page-stack .ant-picker').first());
await shot(ap, '后台-概览-帖子论坛分布', ap.locator('.overview-chart-card', { hasText: '帖子论坛分布' }).first());
await shot(ap, '后台-概览-互动热度分层', ap.locator('.overview-chart-card', { hasText: '互动热度分层' }).first());
await shot(ap, '后台-概览-最新帖子表', ap.locator('.ant-card', { hasText: '最新帖子' }).first());

await goto(ap, '#/forum/forum-list');
await shot(ap, '后台-论坛列表-页面整体');
await shot(ap, '后台-论坛列表-工具栏', ap.locator('.list-table-card .ant-card-head, .list-table-card .card-toolbar, .page-stack .ant-space').first());
await shot(ap, '后台-论坛列表-列表表格', ap.locator('.ant-table').first());
await shot(ap, '后台-论坛列表-行操作', ap.locator('.ant-table-tbody tr').first());

const moreBtn = ap.locator('.ant-table-tbody tr').first().getByRole('button', { name: /更多/ });
if (await moreBtn.count()) {
  await moreBtn.click();
  await ap.waitForTimeout(300);
  await shot(ap, '后台-论坛列表-更多菜单', ap.locator('.ant-dropdown:visible, .ant-dropdown-menu').last());
  await ap.keyboard.press('Escape');
}

const linkBtn = ap.getByRole('button', { name: /查看链接/ }).first();
if (await linkBtn.count()) {
  await linkBtn.click();
  await ap.waitForTimeout(400);
  await shot(ap, '后台-论坛列表-查看链接弹窗', ap.locator('.ant-modal').last());
  await ap.keyboard.press('Escape');
}

await ap.locator('.ant-table-tbody .ant-checkbox-input').first().check({ force: true }).catch(() => {});
await ap.waitForTimeout(200);
await shot(ap, '后台-论坛列表-批量停用', ap.locator('.page-stack').first());
const batchBtn = ap.getByRole('button', { name: '批量停用' });
if (await batchBtn.count()) {
  await batchBtn.click();
  await ap.waitForTimeout(400);
  await shot(ap, '后台-论坛列表-批量停用确认', ap.locator('.ant-modal').last());
  await ap.keyboard.press('Escape');
}

await goto(ap, '#/forum/forum-create');
await shot(ap, '后台-新建论坛-表单页');
await shot(ap, '后台-新建论坛-图标上传', ap.locator('.ant-form-item', { hasText: '论坛图标' }).first());
await shot(ap, '后台-新建论坛-背景图上传', ap.locator('.ant-form-item', { hasText: '背景图' }).first());
await shot(ap, '后台-新建论坛-可见范围', ap.locator('.ant-form-item', { hasText: '可见范围' }).first());
await ap.getByRole('radio', { name: '按部门' }).click();
await ap.waitForTimeout(300);
await shot(ap, '后台-新建论坛-按部门', ap.locator('.ant-form-item', { hasText: '选择部门' }).first());
await ap.getByRole('radio', { name: '自定义人群' }).click();
await ap.waitForTimeout(300);
await shot(ap, '后台-新建论坛-自定义人群', ap.locator('.ant-form-item', { hasText: '选择人员' }).first());
await ap.getByRole('radio', { name: '导入人群' }).click();
await ap.waitForTimeout(300);
await shot(ap, '后台-新建论坛-导入人群', ap.locator('.ant-form-item', { hasText: '导入人群' }).first());
await shot(ap, '后台-新建论坛-底栏操作', ap.locator('.sticky-form-actions').first());

await goto(ap, '#/forum/forum-edit/1');
await shot(ap, '后台-编辑论坛-表单页');

await goto(ap, '#/forum/forum-detail/1');
await shot(ap, '后台-论坛详情-页面整体');
await shot(ap, '后台-论坛详情-标题与操作', ap.locator('.forum-board-title-row').first());
await shot(ap, '后台-论坛详情-基本信息', ap.locator('.ant-descriptions').first());
await shot(ap, '后台-论坛详情-帖子表', ap.locator('.ant-table').first());

const pinMore = ap.locator('.ant-table-tbody tr').first().getByRole('button', { name: /更多/ });
if (await pinMore.count()) {
  await pinMore.click();
  await ap.waitForTimeout(300);
  await shot(ap, '后台-论坛详情-帖子更多菜单', ap.locator('.ant-dropdown:visible').last());
  const pinItem = ap.getByRole('menuitem', { name: /置顶|取消置顶/ }).first();
  if (await pinItem.count()) {
    await pinItem.click();
    await ap.waitForTimeout(400);
    await shot(ap, '后台-论坛详情-置顶确认', ap.locator('.ant-modal').last());
    await ap.keyboard.press('Escape');
  } else {
    await ap.keyboard.press('Escape');
  }
}

const assignBtn = ap.getByRole('button', { name: /指派回复人/ }).first();
if (await assignBtn.count()) {
  await assignBtn.click();
  await ap.waitForTimeout(400);
  await shot(ap, '后台-论坛详情-指派回复人弹窗', ap.locator('.ant-modal').last());
  await ap.keyboard.press('Escape');
}

await goto(ap, '#/forum/topic-detail/1');
await shot(ap, '后台-帖子详情-页面整体');
await shot(ap, '后台-帖子详情-标题与操作', ap.locator('.detail-title-row').first());
await shot(ap, '后台-帖子详情-基本信息', ap.locator('.ant-card', { hasText: '基本信息' }).first());
await shot(ap, '后台-帖子详情-内容与评论', ap.locator('.ant-card', { hasText: '内容' }).first());

const replyTopic = ap.getByRole('button', { name: '回复帖子' });
if (await replyTopic.count()) {
  await replyTopic.click();
  await ap.waitForTimeout(400);
  await shot(ap, '后台-帖子详情-回复帖子表单', ap.locator('.ant-card', { hasText: '内容' }).first());
}

const replyCmt = ap.getByRole('button', { name: /回复评论/ }).first();
if (await replyCmt.count()) {
  await replyCmt.click();
  await ap.waitForTimeout(300);
  await shot(ap, '后台-帖子详情-回复评论', ap.locator('.ant-card', { hasText: '内容' }).first());
}

await goto(ap, '#/forum/topic-detail/1');
const shelfBtn = ap.getByRole('button', { name: /下架|重新上架/ }).first();
if (await shelfBtn.count()) {
  await shelfBtn.click();
  await ap.waitForTimeout(400);
  await shot(ap, '后台-帖子详情-下架确认', ap.locator('.ant-modal').last());
  await ap.keyboard.press('Escape');
}

const topicLink = ap.getByRole('button', { name: '查看链接' });
if (await topicLink.count()) {
  await topicLink.click();
  await ap.waitForTimeout(400);
  await shot(ap, '后台-帖子详情-查看链接弹窗', ap.locator('.ant-modal').last());
  await ap.keyboard.press('Escape');
}

await goto(ap, '#/forum/forum-tags');
await shot(ap, '后台-标签管理-页面整体');
await shot(ap, '后台-标签管理-查询筛选', ap.locator('.search-panel, form, .ant-card').first());
await shot(ap, '后台-标签管理-列表表格', ap.locator('.ant-table').first());
await ap.getByRole('button', { name: '新建标签' }).click();
await ap.waitForTimeout(400);
await shot(ap, '后台-标签管理-新建弹窗', ap.locator('.ant-modal').last());
await ap.keyboard.press('Escape');

const delBtn = ap.getByRole('button', { name: '删除' }).first();
if (await delBtn.count()) {
  await delBtn.click();
  await ap.waitForTimeout(400);
  await shot(ap, '后台-标签管理-删除确认', ap.locator('.ant-modal').last());
  await ap.keyboard.press('Escape');
}

await goto(ap, '#/forum/forum-risk');
await shot(ap, '后台-禁言管理-页面整体');
await shot(ap, '后台-禁言管理-列表表格', ap.locator('.ant-table').first());
await ap.getByRole('button', { name: '添加禁言' }).click();
await ap.waitForTimeout(400);
await shot(ap, '后台-禁言管理-添加弹窗', ap.locator('.ant-modal').last());
await ap.keyboard.press('Escape');

const releaseBtn = ap.getByRole('button', { name: /解除禁言/ }).first();
if (await releaseBtn.count()) {
  await releaseBtn.click();
  await ap.waitForTimeout(400);
  await shot(ap, '后台-禁言管理-解除确认', ap.locator('.ant-modal').last());
  await ap.keyboard.press('Escape');
}

const h5 = await browser.newContext({ viewport: { width: 390, height: 844 }, deviceScaleFactor: 2, isMobile: true, hasTouch: true });
const hp = await h5.newPage();

await goto(hp, '#/c/h5/forum/1', 1100);
await shot(hp, 'H5-板块页-页面整体');
await shot(hp, 'H5-板块页-头图区', hp.locator('.c-forum-hero').first());
await shot(hp, 'H5-板块页-搜索框', hp.locator('.c-forum-search-bar').first());
await shot(hp, 'H5-板块页-标签Tab', hp.locator('.c-forum-tabs').first());
await shot(hp, 'H5-板块页-排序', hp.locator('.c-forum-sort').first());
await shot(hp, 'H5-板块页-帖子列表', hp.locator('.c-forum-posts').first());
await shot(hp, 'H5-板块页-发布按钮', hp.locator('.c-forum-fab').first());

await hp.getByRole('button', { name: '立即发布' }).click();
await hp.waitForTimeout(500);
await shot(hp, 'H5-发布帖子-表单');
await hp.getByRole('button', { name: '关闭' }).click().catch(() => hp.keyboard.press('Escape'));
await hp.waitForTimeout(300);

await goto(hp, '#/c/h5/forum/1?empty=1', 900);
await shot(hp, 'H5-板块页-空态');

await goto(hp, '#/c/h5/forum/999', 800);
await shot(hp, 'H5-板块页-不存在');

await goto(hp, '#/c/h5/forum-topic/1', 1100);
await shot(hp, 'H5-帖子详情-页面整体');
await shot(hp, 'H5-帖子详情-正文', hp.locator('.c-h5-main, .c-forum-topic, main').first());

const likeBtn = hp.getByRole('button', { name: /点赞/ }).first();
if (await likeBtn.count()) await shot(hp, 'H5-帖子详情-点赞收藏', hp.locator('.c-forum-post-foot, .c-forum-topic-actions, footer').first());

const commentReply = hp.getByRole('button', { name: /回复 / }).first();
if (await commentReply.count()) {
  await commentReply.click();
  await hp.waitForTimeout(400);
  await shot(hp, 'H5-帖子详情-评论输入');
}

await goto(hp, '#/c/h5/forum-topic/999', 800);
await shot(hp, 'H5-帖子详情-不存在');

await goto(hp, '#/c/h5/forum-mine', 900);
await shot(hp, 'H5-我的帖子-页面整体');

await goto(hp, '#/c/h5/forum-mine?empty=1', 800);
await shot(hp, 'H5-我的帖子-空态');

const pc = await browser.newContext({ viewport: { width: 1440, height: 900 } });
const pp = await pc.newPage();

await goto(pp, '#/c/pc/forum/1', 1100);
await shot(pp, 'PC-板块页-页面整体');
await shot(pp, 'PC-板块页-头图区', pp.locator('.c-forum-pc-hero').first());
await shot(pp, 'PC-板块页-帖子列表', pp.locator('.c-forum-posts').first());
await shot(pp, 'PC-板块页-侧栏', pp.locator('.c-forum-pc-side').first());
await shot(pp, 'PC-板块页-立即发布', pp.locator('.c-forum-pc-publish').first());
await shot(pp, 'PC-板块页-其他论坛', pp.locator('.c-forum-pc-side-block', { hasText: '其他论坛' }).first());
await shot(pp, 'PC-板块页-热门帖子', pp.locator('.c-forum-pc-side-block', { hasText: '热门帖子' }).first());

await pp.getByRole('button', { name: '立即发布' }).click();
await pp.waitForTimeout(500);
await shot(pp, 'PC-发布帖子-表单', pp.locator('.c-forum-overlay-panel').first());
await pp.getByRole('button', { name: '关闭' }).click().catch(() => pp.keyboard.press('Escape'));

await goto(pp, '#/c/pc/forum/1?empty=1', 900);
await shot(pp, 'PC-板块页-空态');

await goto(pp, '#/c/pc/forum-topic/1', 1100);
await shot(pp, 'PC-帖子详情-页面整体');

await goto(pp, '#/c/pc/forum-mine', 900);
await shot(pp, 'PC-我的帖子-页面整体');

await goto(pp, '#/c', 800);
await shot(pp, 'C端预览-入口页', pp.locator('.c-portal').first());

await writeFile(join(OUT, '_manifest.json'), JSON.stringify(shots, null, 2));
await browser.close();
console.log('done', shots.length);

File: docs/prd/mailbox-app/_capture.py
#!/usr/bin/env python3
"""Capture mailbox PRD screenshots from local Vite."""
from __future__ import annotations

from pathlib import Path

from playwright.sync_api import TimeoutError as PwTimeout
from playwright.sync_api import sync_playwright

ROOT = Path(__file__).resolve().parent / "screenshots"
ROOT.mkdir(parents=True, exist_ok=True)
BASE = "http://127.0.0.1:5173/"
SHOTS: list[str] = []


def save(page, name: str, selector: str | None = None) -> None:
    path = ROOT / f"{name}.png"
    try:
        if selector:
            loc = page.locator(selector).first
            loc.wait_for(state="visible", timeout=4000)
            loc.screenshot(path=str(path), timeout=8000)
        else:
            page.screenshot(path=str(path), full_page=False)
        SHOTS.append(f"{name}.png")
        print("ok", name)
    except Exception as exc:  # noqa: BLE001
        print("fail-el", name, str(exc)[:120])
        page.screenshot(path=str(path), full_page=False)
        SHOTS.append(f"{name}.png")
        print("page", name)


def goto(page, hash_path: str, wait: str = ".ant-layout, .c-h5-shell, .c-pc-shell, .c-portal") -> None:
    page.goto(BASE + hash_path, wait_until="domcontentloaded")
    page.wait_for_timeout(800)
    try:
        page.wait_for_selector(wait, timeout=12000)
    except PwTimeout:
        pass
    page.wait_for_timeout(400)


def admin(page) -> None:
    page.set_viewport_size({"width": 1440, "height": 900})
    goto(page, "#/mailbox/mailbox-overview")
    save(page, "信箱应用-左侧菜单", ".ant-layout-sider")
    save(page, "后台-概览-页面整体")
    save(page, "后台-概览-日期范围", ".overview-date-range")
    save(page, "后台-概览-KPI指标卡", ".overview-kpi-grid")
    save(page, "后台-概览-建言信箱分布", ".overview-chart-card:has-text('建言信箱分布')")
    save(page, "后台-概览-回复情况", ".overview-chart-card:has-text('回复情况')")
    save(page, "后台-概览-待回复建言", ".ant-card:has-text('待回复建言')")

    goto(page, "#/mailbox/mailbox-list")
    save(page, "后台-信箱列表-页面整体")
    save(page, "后台-信箱列表-工具栏", ".ant-card")
    save(page, "后台-信箱列表-列表表格", ".ant-table")
    save(page, "后台-信箱列表-行操作", ".ant-table-tbody tr.ant-table-row")
    more = page.locator(".ant-table-tbody tr.ant-table-row").first.get_by_role("button", name="更多")
    if more.count():
        more.first.click()
        page.wait_for_timeout(300)
        save(page, "后台-信箱列表-更多菜单", ".ant-dropdown")
        stop = page.locator(".ant-dropdown").get_by_text("停用", exact=True)
        if stop.count():
            stop.first.click()
            page.wait_for_timeout(400)
            save(page, "后台-信箱列表-停用确认", ".ant-modal")
            page.keyboard.press("Escape")
            page.wait_for_timeout(300)
        else:
            page.keyboard.press("Escape")

    page.get_by_role("button", name="新建信箱").click()
    page.wait_for_timeout(800)
    save(page, "后台-新建信箱-表单页")
    save(page, "后台-新建信箱-图片与名称", ".edit-form")
    save(page, "后台-新建信箱-可见范围", ".ant-form-item:has-text('可见范围')")
    page.locator(".ant-form-item:has-text('可见范围')").get_by_text("按部门", exact=True).click()
    page.wait_for_timeout(250)
    save(page, "后台-新建信箱-按部门", ".ant-form-item:has-text('选择部门')")
    page.get_by_text("自定义人群", exact=True).click()
    page.wait_for_timeout(250)
    save(page, "后台-新建信箱-自定义人群", ".ant-form-item:has-text('选择人员')")
    page.get_by_text("导入人群", exact=True).click()
    page.wait_for_timeout(250)
    save(page, "后台-新建信箱-导入人群", ".ant-form-item:has-text('导入人群')")
    page.get_by_text("全员", exact=True).click()
    page.locator(".ant-form-item:has-text('响应时效')").get_by_role("switch").click()
    page.wait_for_timeout(250)
    save(page, "后台-新建信箱-响应时效", ".ant-form-item:has-text('响应时效')")
    save(page, "后台-新建信箱-底栏操作", ".sticky-form-actions")
    page.locator(".sticky-form-actions .ant-btn-primary").click(force=True)
    page.wait_for_timeout(400)
    save(page, "后台-新建信箱-校验错误", ".edit-form")

    goto(page, "#/mailbox/mailbox-edit/3")
    save(page, "后台-编辑信箱-表单页")

    goto(page, "#/mailbox/mailbox-detail/3")
    save(page, "后台-信箱详情-页面整体")
    save(page, "后台-信箱详情-标题与操作", ".forum-board-title-row")
    save(page, "后台-信箱详情-信息描述", ".ant-descriptions")
    save(page, "后台-信箱详情-建言列表", ".ant-table")

    goto(page, "#/mailbox/advice-detail/3")
    save(page, "后台-建言详情-已回复-页面整体")
    save(page, "后台-建言详情-基本信息", ".ant-card:has-text('基本信息')")
    save(page, "后台-建言详情-内容", ".ant-card:has-text('内容')")
    save(page, "后台-建言详情-处理人回复", ".ant-card:has-text('处理人回复')")
    save(page, "后台-建言详情-回复建言", ".ant-card:has-text('回复建言')")

    goto(page, "#/mailbox/advice-detail/6")
    save(page, "后台-建言详情-待回复-页面整体")
    save(page, "后台-建言详情-待回复表单", ".moment-inline-reply")

    goto(page, "#/mailbox/advice-detail/9")
    save(page, "后台-建言详情-已驳回种子", ".ant-card:has-text('基本信息')")

    goto(page, "#/mailbox/mailbox-detail/999")
    save(page, "后台-信箱详情-空态", ".ant-empty")


def h5(page) -> None:
    page.set_viewport_size({"width": 390, "height": 844})
    goto(page, "#/c/h5/mailbox", ".c-h5-shell")
    save(page, "H5-信箱列表-页面整体", ".c-h5-frame")
    save(page, "H5-信箱列表-顶栏", ".c-mailbox-hero")
    save(page, "H5-信箱列表-信箱卡片", ".c-mailbox-dirs")
    save(page, "H5-信箱列表-我的建言", ".c-mailbox-mine")
    purposes = page.locator(".c-mailbox-purpose")
    if purposes.count() > 1:
        purposes.nth(1).click()
        page.wait_for_timeout(250)
        save(page, "H5-信箱列表-简介展开", ".c-mailbox-dir")
    page.get_by_role("button", name="提交 战略发展建言").first.click()
    page.wait_for_timeout(500)
    save(page, "H5-提交建言-弹层", ".c-forum-overlay-panel")
    page.locator(".c-forum-submit").click()
    page.wait_for_timeout(400)
    save(page, "H5-提交建言-标题校验", ".c-h5-shell")
    page.locator(".c-forum-overlay-scrim").click(force=True)
    page.wait_for_timeout(200)

    goto(page, "#/c/h5/mailbox?empty=1", ".c-h5-shell")
    save(page, "H5-信箱列表-空建言", ".c-mailbox-mine")

    goto(page, "#/c/h5/mailbox/22", ".c-h5-shell")
    save(page, "H5-建言详情-页面整体", ".c-h5-frame")
    save(page, "H5-建言详情-内容卡片", ".c-mailbox-detail-card")

    goto(page, "#/c/h5/mailbox/999", ".c-h5-shell")
    save(page, "H5-建言详情-不存在", ".c-mailbox-empty")


def pc(page) -> None:
    page.set_viewport_size({"width": 1440, "height": 900})
    goto(page, "#/c/pc/mailbox", ".c-pc-shell")
    save(page, "PC-信箱列表-页面整体")
    save(page, "PC-信箱列表-顶栏", ".c-pc-header")
    save(page, "PC-信箱列表-信箱卡片", ".c-mailbox-dirs")
    save(page, "PC-信箱列表-我的建言", ".c-mailbox-mine")
    page.get_by_role("button", name="提交 员工体验建言").first.click()
    page.wait_for_timeout(500)
    save(page, "PC-提交建言-弹层", ".c-forum-overlay-panel")
    page.locator(".c-forum-overlay-scrim").click(force=True)

    goto(page, "#/c/pc/mailbox?empty=1", ".c-pc-shell")
    save(page, "PC-信箱列表-空建言", ".c-mailbox-mine")

    goto(page, "#/c/pc/mailbox/23", ".c-pc-shell")
    save(page, "PC-建言详情-页面整体")
    save(page, "PC-建言详情-内容卡片", ".c-mailbox-detail-card")
    save(page, "PC-建言详情-返回列表", ".c-mailbox-pc-back")

    goto(page, "#/c/pc/mailbox/999", ".c-pc-shell")
    save(page, "PC-建言详情-不存在", ".c-mailbox-empty")

    goto(page, "#/c", ".c-portal")
    save(page, "C端预览-入口页")


def main() -> None:
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True, channel="chrome")
        page = browser.new_page()
        admin(page)
        h5(page)
        pc(page)
        browser.close()
    (ROOT / "_manifest.json").write_text("[\n  " + ",\n  ".join(f'"{n}"' for n in SHOTS) + "\n]\n", encoding="utf-8")
    print("done", len(SHOTS))


if __name__ == "__main__":
    main()
